### PR TITLE
added ability to get `Res<T>` from `World` with `World::get_resource_ref`

### PR DESCRIPTION
**Pull request by atornity**
# Objective

It's sometimes desirable to get a `Res<T>` rather than `&T` from `World::get_resource`.
Alternative to #9940, partly adresses #9926

## Solution

added additional methods to `World` and `UnsafeWorldCell` to retrieve a resource wrapped in a `Res`.
- `UnsafeWorldCell::get_resource_ref`
- `World::get_resource_ref`
- `World::resource_ref`

I can change it so `World::resource_mut` returns `ResMut` instead of `Mut` as well if that's desired, but that could also be added later in a seperate pr.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> I can change it so World::resource_mut returns ResMut instead of Mut as well if that's desired, but that could also be added later in a seperate pr.

Let's keep this split this out.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -1335,6 +1359,15 @@ impl World {
         unsafe { self.as_unsafe_world_cell_readonly().get_resource() }
     }
 
+    /// Gets a reference to the resource of the given type if it exists
```
Comment (hymm) [inline crates/bevy_ecs/src/world/mod.rs:1362]: This is the same doc as is on `get_resource`. We should make the distinction clear between the two.

```suggestion
    /// Gets a reference including change detection to the resource of the given type if it exists.
```

File: crates/bevy_ecs/src/world/unsafe_world_cell.rs [hunk 2]
```diff
@@ -342,6 +342,30 @@ impl<'w> UnsafeWorldCell<'w> {
         }
     }
 
+    /// Gets a reference to the resource of the given type if it exists
```
Comment (hymm) [inline crates/bevy_ecs/src/world/unsafe_world_cell.rs:345]:
```suggestion
    /// Gets a reference including change detection to the resource of the given type if it exists.
```

File: crates/bevy_ecs/src/world/unsafe_world_cell.rs [hunk 3]
```diff
@@ -342,6 +342,30 @@ impl<'w> UnsafeWorldCell<'w> {
...
+    /// # Safety
+    /// It is the callers responsibility to ensure that
+    /// - the [`UnsafeWorldCell`] has permission to access the resource
+    /// - no mutable reference to the resource exists at the same time
+    #[inline]
+    pub unsafe fn get_resource_ref<R: Resource>(self) -> Option<Res<'w, R>> {
+        let component_id = self.components().get_resource_id(TypeId::of::<R>())?;
+
+        // SAFETY: caller ensures `self` has permission to access the resource
+        //  caller also ensure that no mutable reference to the resource exists
```
Comment (hymm) [inline crates/bevy_ecs/src/world/unsafe_world_cell.rs:356]:
```suggestion
        // caller also ensure that no mutable reference to the resource exists
```

**Review by hymm — APPROVED** [review]
Looks good. Just some minor comments on wording.

**Comment by alice-i-cecile** [comment]
[CI failure](https://github.com/bevyengine/bevy/actions/runs/7680715381/job/20933085710?pr=11561) is real, but doesn't seem related to your PR.

>     --> crates/bevy_ecs/src/world/mod.rs:1911:74
     |
1911 |     /// invalidating all [`Entity`] and resource fetches such as [`Res`](crate::system::Res), [`ResMut`](crate::system::ResMut)
     |                                                                   -----  ^^^^^^^^^^^^^^^^^^ explicit target is redundant
     |                                                                   |
     |                                                                   because label contains path that resolves to same destination

**Comment by mockersf** [comment]
> [CI failure](https://github.com/bevyengine/bevy/actions/runs/7680715381/job/20933085710?pr=11561) is real, but doesn't seem related to your PR.

it is? `Res` is now imported, so ci is telling that the explicit path is not needed anymore

**Comment by doonv** [comment]
Why a `Res<T>` instead of a `Ref<T>`?

**Comment by atornity** [comment]
> > [CI failure](https://github.com/bevyengine/bevy/actions/runs/7680715381/job/20933085710?pr=11561) is real, but doesn't seem related to your PR.
> 
> it is? `Res` is now imported by this PR, so CI is telling that the explicit path is not needed anymore

yup, should be fixed now

File: crates/bevy_ecs/src/world/mod.rs [hunk 4]
```diff
@@ -1302,6 +1302,30 @@ impl World {
         }
     }
 
+    /// Gets a reference to the resource of the given type
+    ///
+    /// # Panics
+    ///
+    /// Panics if the resource does not exist.
+    /// Use [`get_resource`](World::get_resource) instead if you want to handle this case.
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/world/mod.rs:1310]:
```suggestion
    /// Use [`get_resource_ref`](World::get_resource_ref) instead if you want to handle this case.
```

Comment (atornity) [reply crates/bevy_ecs/src/world/mod.rs]: whops, thanks

**Pull request merged**